[Ingestion pipeline] Ingest observations without GroupByKey

rohitkumarbhagat opened the pull request:
Changes
- Use GBK only for graph mutations. Observations can be written directly.
- Limit default sharding to be 1. This helps in co-locating all rows for a subject-id in same bundle -> same spanner batch -> less splits per batch
- Filter duplicate edges before and after GBK
- Update spanner IO config for better throughput

keyurva commented on pipeline/ingestion/src/main/java/org/datacommons/ingestion/pipeline/Transforms.java:
```diff
@@ -148,20 +172,57 @@ public ImportGroupTransform(
+      return groupByGraphOnly(cacheRows);
     }
 
     private PCollection<Void> groupBy(PCollection<String> cacheRows) {
```
If this is not used anymore, we can remove it.

rohitkumarbhagat replied:
Sure.

keyurva commented on pipeline/ingestion/src/main/java/org/datacommons/ingestion/pipeline/Transforms.java:
```diff
@@ -148,20 +172,57 @@ public ImportGroupTransform(
+
+      var observations =
+          kvs.get(observationTag).apply("ExtractObservationMutations", Values.create());
+      // TODO: Explore emitting protos instead of mutations to reduce shuffle cost.
```
This is a good point. Should try this in a future PR. The shuffle is one of our most expensive steps, should this should help.

keyurva commented on pipeline/ingestion/src/main/java/org/datacommons/ingestion/pipeline/Transforms.java:
```diff
@@ -26,49 +26,60 @@ static class CacheRowKVMutationsDoFn extends DoFn<String, KV<String, Mutation>>
     private final SpannerClient spannerClient;
     private final SkipProcessing skipProcessing;
+    private final TupleTag<KV<String, Mutation>> graphTag;
+    private final TupleTag<KV<String, Mutation>> observationTag;
```
Since we're not grouping obs, is there a benefit to emitting KV pairs for obs?

rohitkumarbhagat replied:
You are right. Observation keys are not used. However, it does not add to much cost (mostly key construction). I would still like to keep it as it might help while trying out different pipeline configs later.

keyurva commented on pipeline/ingestion/src/main/java/org/datacommons/ingestion/pipeline/Transforms.java:
```diff
@@ -148,20 +172,57 @@ public ImportGroupTransform(
+      var write =
+          PCollectionList.of(graph)
+              .and(observations)
+              .apply("MergeMutations", Flatten.<Mutation>pCollections())
```
Does flattening cause any delay in processing obs mutations? I'm assuming since there's no group by for obs, they must get to this stage much sooner than the graph mutations.

rohitkumarbhagat replied:
Flattening does cause any delay. It emits as it gets.

keyurva commented on pipeline/ingestion/src/main/java/org/datacommons/ingestion/spanner/SpannerClient.java:
```diff
@@ -42,14 +41,15 @@ private SpannerClient(Builder builder) {
-        .withBatchSizeBytes(3 * 1024 * 1024)
-        .withMaxNumMutations(10000)
-        .withGroupingFactor(100)
+        .withBatchSizeBytes(500 * 1024) // decrease batch size for observations (bigger rows)
```
Awesome reasoning and comments - thanks!

keyurva replied:
nit: consider assigning them to consts to make the comments more explicit.

rohitkumarbhagat replied:
Good idea.

keyurva commented on pipeline/ingestion/src/main/java/org/datacommons/ingestion/spanner/SpannerClient.java:
```diff
@@ -138,17 +138,30 @@ public List<KV<String, Mutation>> filterObservationKVMutations(
           continue;
         }
         seenNodes.add(subjectId);
+      } else if (seenEdges != null && mutation.getTable().equals(edgeTableName)) {
```
nit: when would `seenEdges` be null?

rohitkumarbhagat replied:
Removed.

keyurva reviewed (approved):
Thanks Rohit - this is great!

I have a couple of comments regarding KV pairs for obs and flattening. I'll approve so as to unblock you but pl address those 2 comments.

rohitkumarbhagat commented:
Thanks Keyur for the approval. Auto-submit was enabled, hence PR got merged. Apologies. I will address the comments in next PR.

rohitkumarbhagat commented on pipeline/ingestion/src/main/java/org/datacommons/ingestion/pipeline/Transforms.java:
```diff
@@ -26,49 +26,60 @@ static class CacheRowKVMutationsDoFn extends DoFn<String, KV<String, Mutation>>
         Metrics.counter(CacheRowKVMutationsDoFn.class, "dc_duplicate_obs_creation");
     private static final Counter DUPLICATE_NODES_COUNTER =
         Metrics.counter(CacheRowKVMutationsDoFn.class, "dc_duplicate_nodes_creation");
+    private static final Counter DUPLICATE_EDGES_COUNTER =
```
Max: 6M for disaster import group
<img width="603" alt="image" src="https://github.com/user-attachments/assets/d09ca09f-9190-486e-8b56-1e2cc1953312" />

rohitkumarbhagat commented on pipeline/ingestion/README.md:
````diff
@@ -11,28 +11,29 @@ Example usages:
 ```shell
 mvn -Pdataflow-runner compile exec:java -pl ingestion -am -Dexec.mainClass=org.datacommons.ingestion.pipeline.IngestionPipeline \
--Dexec.args="--project=datcom-store --gcpTempLocation=gs://keyurs-dataflow/temp --runner=DataflowRunner --region=us-central1  --numWorkers=50 --dataflowServiceOptions=enable_google_cloud_profiler --workerMachineType=e2-highmem-16"
+-Dexec.args="--project=datcom-store --gcpTempLocation=gs://keyurs-dataflow/temp --runner=DataflowRunner --region=us-central1  --numWorkers=60 --maxNumWorkers=60 --dataflowServiceOptions=enable_google_cloud_profiler --workerMachineType=n2-custom-16-262144-ext"
````
Will add in a separate section

rohitkumarbhagat commented:
> Thanks Rohit - this is great!
> 
> I have a couple of comments regarding KV pairs for obs and flattening. I'll approve so as to unblock you but pl address those 2 comments.

Addressed comments in https://github.com/datacommonsorg/import/pull/382
